File: src/main/java/seremis/geninfusion/soul/Allele.java
package seremis.geninfusion.soul;

import net.minecraft.nbt.NBTTagCompound;
import seremis.geninfusion.api.soul.IAllele;
import seremis.geninfusion.api.soul.EnumAlleleType;

import java.lang.reflect.Constructor;

public class Allele implements IAllele {

    public boolean isDominant;
    public EnumAlleleType type;
    
    public Allele(boolean isDominant, EnumAlleleType type) {
        this.isDominant = isDominant;
        this.type = type;
    }

    public Allele(Object... args) {
        isDominant = (Boolean) args[0];
    }
    
    public Allele(NBTTagCompound compound) {
        readFromNBT(compound);
    }
    
    @Override
    public boolean isDominant() {
        return isDominant;
    }

    @Override
    public void writeToNBT(NBTTagCompound compound) {
        compound.setBoolean("isDominant", isDominant);
        compound.setInteger("type", type.ordinal());
    }

    @Override
    public void readFromNBT(NBTTagCompound compound) {
        isDominant = compound.getBoolean("isDominant");
        type = EnumAlleleType.values()[compound.getInteger("type")];
    }
    
    public static IAllele readAlleleFromNBT(NBTTagCompound compound) {
        EnumAlleleType type = EnumAlleleType.values()[compound.getInteger("type")];
        
        try {
            Constructor<?> ctor = type.clazz.getConstructor(NBTTagCompound.class);
            Object object = ctor.newInstance(compound);
            
            return (IAllele) object;
        } catch(Exception e) {
            e.printStackTrace();
        }
        
        return null;
    }
    
    @Override
    public String toString() {
    	return "Allele[type: " + type + ", isDominant: " + isDominant + "]";
    }
}

File: src/main/scala/seremis/geninfusion/soul/Allele.scala
package seremis.geninfusion.soul

import net.minecraft.nbt.NBTTagCompound
import seremis.geninfusion.api.soul.{IAllele, IAlleleType, SoulHelper}

class Allele(var dominant: Boolean, var alleleData: Any, var alleleType: IAlleleType) extends IAllele {

    override def isDominant: Boolean = dominant

    override def getAlleleData: Any = alleleData
    override def setAlleleData(data: Any) = alleleData = data
    override def getAlleleType: IAlleleType = alleleType

    def this(dominant: Boolean, alleleData: Any, alleleType: Class[_]) {
        this(dominant, alleleData, SoulHelper.alleleTypeRegistry.getAlleleTypeForClass(alleleType))
    }

    def this(compound: NBTTagCompound) {
        this(false, false, AlleleType.typeBoolean)
        dominant = compound.getBoolean("dominant")
        alleleType = SoulHelper.alleleTypeRegistry.getAlleleTypeFromId(compound.getInteger("alleleType"))
        alleleData = alleleType.readFromNBT(compound, "alleleData")
    }

    override def writeToNBT(compound: NBTTagCompound): NBTTagCompound = {
        compound.setInteger("alleleType", SoulHelper.alleleTypeRegistry.getAlleleTypeId(alleleType))
        if(alleleData != null)
            alleleType.writeToNBT(compound, "alleleData", alleleData)
        compound.setBoolean("dominant", dominant)
        compound
    }

    override def readFromNBT(compound: NBTTagCompound): NBTTagCompound = {
        alleleType = SoulHelper.alleleTypeRegistry.getAlleleTypeFromId(compound.getInteger("alleleType"))
        if(compound.hasKey("alleleData"))
            alleleData = alleleType.readFromNBT(compound, "alleleData")
        dominant = compound.getBoolean("dominant")
        compound
    }

    override def toString: String = {
        "Allele:[isDominant: " + isDominant + ", alleleType: " + alleleType + ", alleleData: " + alleleData + "]"
    }
}